nexus-and-alo/nexus-strawman.md

```diff
@@ -0,0 +1,361 @@
+**Contents:**
+
+- [Terminology](#terminology)
+- [User flows](#user-flows)
+  - [Hosting a Nexus Service](#hosting-a-nexus-service)
+  - [Consuming a Nexus Service from Temporal](#consuming-a-nexus-service-from-temporal)
+- [Concepts](#concepts)
+  - [Nexus Instantiation](#nexus-instantiation)
+  - [Nexus CRUD](#nexus-crud)
+  - [Nexus Task hosting](#nexus-task-hosting)
+  - [Nexus Task Worker](#nexus-task-worker)
+  - [Calling Nexus from Temporal](#calling-nexus-from-temporal)
+
+## Terminology
+
+**Nexus Service Instance:** Formally registered Nexus Service with a Nexus backend (eg: Temporal).  Nexus Instances must always have a unique identifier.
+
+**Nexus Task:** Unit of work which is run as a result of a request reaching the Nexus endpoint. Nexus Tasks are developed and operated similar to WorkflowTasks and ActivityTasks (eg: run and hosted on Nexus Workers). Within a Nexus Task Handler, developers can transform incoming payloads, make routing decisions and more.
+
+**Nexus Task Handler:** Source code which is designed to handle incoming Nexus Tasks. Nexus Task Handler is similar to a Workflow or Activity definition and is hosted by a Nexus Worker. 
+
+**Nexus Worker:** A special purpose Worker that pulls Nexus Tasks for its registered Nexus Instances from the Server and executes them using associated Nexus Task Handlers. 
```

Comment:
Users might want to consume nexus requests using standard a HTTP / gRPC stack.
I would note that that conversion of requests to tasks just one implementation option.

Reply:
Mostly agree with chad on this one.

Reply:
@cretz then why would you suggest calling "Nexus Task" a "Nexus Request"?

Part of the objectives of this proposal is to make a non-Temporal specific standard and to allow Nexus requests to be sent over open protocols like HTTP.

Reply:
It is a non-Temporal specific standard, and Nexus requests (one day soon) will be able to be invoked over HTTP. But what you're referring to is _request handling_, not _request sending_. I think we all agree request sending needs to be HTTP, but I question whether the language that Nexus workers speak to the backend to handle requests needs to be abstracted over HTTP.

Reply:
Nexus requests can be sent over temporal-specific, or other protocols (like HTTP and gRPC).

The request handler can be implemented using standard HTTP and gRPC stack or when Temporal is the gateway using a Temporal Nexus worker.
I see value in designing a generic, modular Nexus SDK API to enable reuse of most of the handler code in both worker and HTTP / gRPC scenarios.

There are a few layers here on the request handling (backend) side:
1. gateway
2. request -> task dispatcher
3. task -> handler dispatcher
4. handler

Layer 1 can be any HTTP / gRPC gateway or Temporal.
Layer 2 is only required when using a non-standard protocol and will be implemented by Temporal server. Alternative implementations are allowed of course (we'll probably end up with a test server dispatcher).
Layer 3 is only relevant if using layer 2.
Layer 4 can be used on top of layer 3, or layer 1 with an additional thin adapter layer.

Reply:
I like the idea of being able to set up a Nexus directly backed by serverless functions (1 -> 4, Requests are sent from gateway to function over HTTP)